examples/core/dof/app.js

```diff
@@ -0,0 +1,433 @@
+
+
+    //////////////////////////////////////////
+    // Initial transforms for instanced boxes
+    //////////////////////////////////////////
+
+    let boxXforms = new Array(NUM_BOXES);
+    let boxMatrices = new Float32Array(NUM_BOXES * 16);
+    let boxI = 0;
+    for (let j = 0; j < NUM_ROWS; ++j) {
+        let rowOffset = (j - Math.floor(NUM_ROWS / 2));
+        for (let i = 0; i < BOXES_PER_ROW; ++i) {
+          let scale = [0.4, 0.4, 0.4];
+          let rotate = [-Math.random() * Math.PI, 0, Math.random() * Math.PI];
+          let translate = [-i + 2 - rowOffset, 0, -i + 2 + rowOffset];
+          boxXforms[boxI] = {
+            scale: scale,
+            translate: translate,
+            rotate: rotate,
+            matrix: new Matrix4().translate(translate).rotateXYZ(rotate).scale(scale)
+          };
+            
+          boxMatrices.set(boxXforms[boxI].matrix, boxI * 16);
+        ++boxI;
+        }
+    }
+
+
+    ////////////////////////////////////////////////
+    // Set up frambuffers. 
+    ////////////////////////////////////////////////
+
+    // Need to ensure both color and depth targets can be sampled.
+    const mainFramebuffer = new Framebuffer(gl, {
+      width: gl.drawingBufferWidth,
+      height: gl.drawingBufferHeight,
+      attachments: {
+        [GL.COLOR_ATTACHMENT0]: new Texture2D(gl, {
+          format: GL.RGBA,
+          type: GL.UNSIGNED_BYTE,
+          width: gl.drawingBufferWidth,
+          height: gl.drawingBufferHeight,
+          mipmaps: false,
+          parameters: {
+            [GL.TEXTURE_MIN_FILTER]: GL.LINEAR,
+            [GL.TEXTURE_MAG_FILTER]: GL.LINEAR,
+            [GL.TEXTURE_WRAP_S]: GL.CLAMP_TO_EDGE,
+            [GL.TEXTURE_WRAP_T]: GL.CLAMP_TO_EDGE
+          }
+        }),
+        [GL.DEPTH_ATTACHMENT]: new Texture2D(gl, {
+          format: GL.DEPTH_COMPONENT16,
+          type: GL.UNSIGNED_SHORT,
+          dataFormat: GL.DEPTH_COMPONENT,
+          width: gl.drawingBufferWidth,
+          height: gl.drawingBufferHeight,
+          mipmaps: false,
+          parameters: {
+            [GL.TEXTURE_MIN_FILTER]: GL.NEAREST,
+            [GL.TEXTURE_MAG_FILTER]: GL.NEAREST,
+            [GL.TEXTURE_WRAP_S]: GL.CLAMP_TO_EDGE,
+            [GL.TEXTURE_WRAP_T]: GL.CLAMP_TO_EDGE
+          }
+        })
+      }
+    })
+
+    // Postprocessing FBO doesn't need a depth attachment.
+    const dofFramebuffer = new Framebuffer(gl, { width: gl.drawingBufferWidth, height: gl.drawingBufferHeight, depth: false});
+
+    return loadTextures(gl, {
+      urls: ['webgl-logo.png'],
+      mipmaps: true,
+      parameters: {
+        [gl.TEXTURE_MAG_FILTER]: gl.LINEAR,
+        [gl.TEXTURE_MIN_FILTER]: gl.LINEAR_MIPMAP_NEAREST
+      }
+    }).then((textures) => {
+
+      ////////////////////////////////////////////
+      // Set up instanced model to draw boxes.
+      ////////////////////////////////////////////
+
+      const cubeGeo = new CubeGeometry().getAttributes();
```

Comment:
Instead of extracting all these details, can you subclass from `Cube` and add your customizations there? Like this : https://github.com/uber/luma.gl/blob/05263ca5245cffced569bc7e9c8d63ad1e3c78a2/examples/core/instancing/app.js#L16